gradio/components/audio.py

```diff
@@ -102,13 +105,11 @@ def __init__(
             show_download_button: If True, will show a download button in the corner of the component for saving audio. If False, icon does not appear.
             show_share_button: If True, will show a share icon in the corner of the component that allows user to share outputs to Hugging Face Spaces Discussions. If False, icon does not appear. If set to None (default behavior), then the icon appears if this Gradio app is launched on Spaces, but not otherwise.
             show_edit_button: If True, will show an edit icon in the corner of the component that allows user to edit the audio. If False, icon does not appear. Default is True.
+            min_length: The minimum length of audio (in seconds) that the user can pass into the prediction function. If None, there is no minimum length.
+            max_length: The maximum length of audio (in seconds) that the user can pass into the prediction function. If None, there is no maximum length.
+            waveform_options: A dictionary of options for the waveform display. Options include: waveform_color (str), waveform_progress_color (str), show_controls (bool). Default is None, which uses the default values for these options.
         """
-        valid_sources = ["upload", "microphone"]
-        source = source if source else ("microphone" if streaming else "upload")
-        if source not in valid_sources:
-            raise ValueError(
-                f"Invalid value for parameter `source`: {source}. Please choose from one of: {valid_sources}"
-            )
+        source = list(set(source)) if source is not None else (["microphone"] if streaming else ["microphone", "upload"])
```

Comment:
I like that you convert strings to lists, but just wondering, do you need the `set` here? Or could you just do:
```suggestion
        source = list(source) if source is not None else (["microphone"] if streaming else ["microphone", "upload"])
```

Reply:
Ideally if a user accidentally passes in a string (e.g. "upload"), it should treat it as the same as ["upload"], which `list(source)` does, but `list(set(source))` does not